### ISSUE TITLE
`which echo` is misleading

### ISSUE THREAD

**Issue by l0b0** [missing-semester/missing-semester#12]
The [overview page](https://missing.csail.mit.edu/2020/course-shell/) uses `which echo` to show which program is executed when running `echo $PATH`. This is misleading. `which` only shows executables on the `$PATH`, but `echo` is built into most shells. For example, in Bash:

```
$ type -a echo
echo is a shell builtin
echo is /bin/echo
```

So when I run `echo $PATH` (or [preferably](https://mywiki.wooledge.org/Quotes) `echo "$PATH"`) it does *not* run `/bin/echo`, but rather the shell built-in. This is also mentioned in the GNU coreutils `man bash`:

> NOTE:  your  shell  may have its own version of echo, which usually supersedes the version described here.  Please refer to your shell's documentation for details about the options it supports.

**Comment by aGIToz** [comment]
Its true! Just verified it by,
```bash
$ sudo chmod a-x /bin/echo
$ echo "TEST"
TEST
```
Which means it doesn't run the /bin/echo

**Comment by jonhoo** [comment]
While it is true that `echo` is often a shell built-in, there is no fundamental reason why it _needs_ to be. All the examples we given work just as well whether you use the built-in `echo` or `/bin/echo`. Now, it could be that it'd be good to talk about shell built-ins (like `cd`), but I think it's fine to pretend that `echo` is a standalone program the purposes of explanation here. I can't immediately think of a non-built-in program that serves as well in examples as `echo` does.

**Comment by anishathalye** [comment]
I agree with @jonhoo. I also understand that this example might be confusing. Perhaps we should add a brief explanation of shell builtins, or at least provide a link. Feel free to submit a PR, otherwise I'll get around to adding this eventually. Let's keep the issue open until this is addressed.

**Comment by l0b0** [comment]
It's been problematic for me at least once in the last couple decades, when the built-in and `$PATH` `echo`s functioned differently (I believe only one of them supported `-e` or something). I've no idea how common it is for them to be different, but a simple workaround would be to reference something popular which I believe is almost never a built-in, like `grep`, `find`, `man`, `locate`, `git`, etc.

**Comment by anishathalye** [comment]
One nice thing about `echo` is that it's functionality is obvious based on its name (or at least more obvious than tools like `grep`). Are there any other non-builtin commands that have relatively obvious functionality?

**Comment by bewuethr** [comment]
I just watched/read the first lecture and noticed this as well. Maybe mentioning it in a footnote might be worthwhile? I understand that simplifications are necessary to keep the scope manageable, of course.

@l0b0 Regarding how common it is for different `echo` implementations to do different things: there's [this page](https://www.in-ulm.de/~mascheck/various/echo+printf/) that goes into great detail of differences, and the [POSIX spec for `echo`](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/echo.html) recommends using `printf` instead fo portability (under "Application usage").

**Comment by l0b0** [comment]
@bewuethr Yeah, I use `printf` for basically anything that's not a static string, especially the `%q` Bashism is quite useful if you want to *safely* print some user input to screen.
